### PR TITLE
Source Metabase: migrate to Beta YAML

### LINKED ISSUE #18706: Source Metabase: Migrate connector from Alpha (Python) to Beta (YAML)
Session Token used: https://www.metabase.com/learn/administration/metabase-api

### PR DESCRIPTION
**Pull request by darynaishchenko**
source Metabase: [check list](https://docs.google.com/spreadsheets/d/1W1w98ZcIwHWqK0HuqPjIqTKsyQLmD0757nA1IENEmTY/edit#gid=1288588270)

### CODE REVIEW

File: airbyte-integrations/connectors/source-metabase/source_metabase/source.py [hunk 1]
```diff
@@ -3,117 +3,35 @@
...
-from airbyte_cdk.sources.streams.http.auth import HttpAuthenticator
-from source_metabase.streams import Activity, Cards, Collections, Dashboards, Users
+from airbyte_cdk.sources.declarative.yaml_declarative_source import YamlDeclarativeSource
+
+from .components import MetabaseAuth
 
 API_URL = "instance_api_url"
 USERNAME = "username"
 PASSWORD = "password"
 SESSION_TOKEN = "session_token"
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-metabase/source_metabase/source.py:16]: I believe this should have been removed

File: airbyte-integrations/connectors/source-metabase/source_metabase/source.py [hunk 2]
```diff
@@ -3,117 +3,35 @@
...
+            session = MetabaseAuth(config, {})
             return session.has_valid_token(), None
         except Exception as e:
             return False, e
         finally:
             if session:
                 session.close_session()
 
-    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
-        self.session = MetabaseAuth(logging.getLogger("airbyte"), config)
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-metabase/source_metabase/source.py:105]: `self.close_session()` will never execute since `self.session` is never set

File: airbyte-integrations/connectors/source-metabase/source_metabase/source.py [hunk 3]
```diff
@@ -3,117 +3,35 @@
...
-class SourceMetabase(AbstractSource):
+class SourceMetabase(YamlDeclarativeSource):
     def __init__(self):
         self.session = None
+        super().__init__(**{"path_to_yaml": "metabase.yaml"})
 
     def check_connection(self, logger: logging.Logger, config: Mapping[str, Any]) -> Tuple[bool, Any]:
         session = None
         try:
-            session = MetabaseAuth(logger, config)
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-metabase/source_metabase/source.py:96]: On one hand you are defining a check in yaml with a stream passed in, on the other hand you're overriding a `check` method in python. This means yaml check will not work and has no sense. We have to leave just one of them, I'm in favour of yaml based check

File: airbyte-integrations/connectors/source-metabase/source_metabase/components.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                f"Connection check for Metabase successful for {json_response['common_name']} login at {json_response['last_login']}"
+            )
+            return True
+        else:
+            raise ConnectionError(f"Failed to retrieve new session token, response code {response.status_code} because {response.reason}")
+
+    def get_auth_header(self) -> Mapping[str, Any]:
+        return {"X-Metabase-Session": self.session_token}
+
+    def close_session(self):
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-metabase/source_metabase/components.py:73]: is this ever called?

**Review by davydov-d — CHANGES_REQUESTED** [review]

File: airbyte-integrations/connectors/source-metabase/source_metabase/components.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+API_URL_CONFIG_KEY = "instance_api_url"
+USERNAME_CONFIG_KEY = "username"
+PASSWORD_CONFIG_KEY = "password"
+SESSION_TOKEN_CONFIG_KEY = "session_token"
+
+
+@dataclass
+class MetabaseAuth(HttpAuthenticator, JsonSchemaMixin):
+    def __init__(self, config: Mapping[str, Any], options: Mapping[str, Any]):
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-metabase/source_metabase/components.py:21]: One more thing: we used to have one session per sync, now we will have one session per stream. I suggest you make it a source instance attribute/cached property/etc

File: airbyte-integrations/connectors/source-metabase/acceptance-test-config.yml [hunk 6]
```diff
@@ -1,20 +1,26 @@
...
+        backward_compatibility_tests_config:
+          disable_for_version: "0.1.0"
   basic_read:
-    - config_path: "secrets/config.json"
-      configured_catalog_path: "integration_tests/configured_catalog.json"
+    tests:
+      - config_path: "secrets/config.json"
+        configured_catalog_path: "integration_tests/configured_catalog.json"
+        expect_records:
+          path: "integration_tests/expected_records.txt"
```
[the remark this answers is not in the document]
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/acceptance-test-config.yml:22]: added

File: airbyte-integrations/connectors/source-metabase/acceptance-test-config.yml [hunk 7]
```diff
@@ -1,20 +1,26 @@
...
-tests:
+acceptance_tests:
   spec:
-    - spec_path: "source_metabase/spec.yaml"
-      backward_compatibility_tests_config:
-        disable_for_version: "0.1.0"
+    tests:
+      - spec_path: "source_metabase/spec.yaml"
+        backward_compatibility_tests_config:
+          disable_for_version: "0.1.0"
```
[the remark this answers is not in the document]
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/acceptance-test-config.yml]: removed

File: airbyte-integrations/connectors/source-metabase/source_metabase/metabase.yaml [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+      $ref: "*ref(definitions.data_field_selector)"
+    paginator:
+      type: NoPagination
+    requester:
+      $ref: "*ref(definitions.requester)"
+  base_stream:
+    primary_key: "id"
+    retriever:
+      $ref: "*ref(definitions.retriever)"
+  data_field_base_stream:
```
[the remark this answers is not in the document]
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/source_metabase/metabase.yaml]: done

File: airbyte-integrations/connectors/source-metabase/source_metabase/source.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/source_metabase/source.py:17]: done

File: airbyte-integrations/connectors/source-metabase/source_metabase/source.py (hunk 2 above)
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/source_metabase/source.py:105]: fixed

File: airbyte-integrations/connectors/source-metabase/source_metabase/source.py (hunk 3 above)
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/source_metabase/source.py:96]: thanks, we need custom check because of updating session token, so I left check here, and removed from yaml.

File: airbyte-integrations/connectors/source-metabase/source_metabase/components.py (hunk 4 above)
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/source_metabase/components.py:73]: yes

File: airbyte-integrations/connectors/source-metabase/source_metabase/components.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+from dataclasses_jsonschema import JsonSchemaMixin
+
+API_URL_CONFIG_KEY = "instance_api_url"
+USERNAME_CONFIG_KEY = "username"
+PASSWORD_CONFIG_KEY = "password"
+SESSION_TOKEN_CONFIG_KEY = "session_token"
+
+
+@dataclass
+class MetabaseAuth(HttpAuthenticator, JsonSchemaMixin):
```
[the remark this answers is not in the document]
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/source_metabase/components.py:20]: done

File: airbyte-integrations/connectors/source-metabase/source_metabase/source.py [hunk 10]
```diff
@@ -102,17 +34,9 @@ def check_connection(self, logger: logging.Logger, config: Mapping[str, Any]) ->
...
-        return [
-            Activity(**args),
-            Cards(**args),
-            Collections(**args),
-            Dashboards(**args),
-            Users(**args),
-        ]
+        self.session = MetabaseAuth(config, {})
+
+        return super().streams(config)
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-metabase/source_metabase/source.py:39]: Once again: you're not passing session object here. This means new session will be instantiated per each stream if an authenticator is passed to a stream in yaml. After the sync succeeds, only the source session (defined two lines above) will be closed, but not those sessions in the streams, right?

**Review by girarda — CHANGES_REQUESTED** [review]
this is great! Only request is to move the authenticator to the CDK's airbyte_cdk/sources/declarative/auth package so it can be reused. Will approve when it's done

File: airbyte-integrations/connectors/source-metabase/source_metabase/metabase.yaml [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "collections"
+      path: "collection"
+  dashboards_stream:
+    $ref: "*ref(definitions.base_stream)"
+    $options:
+      name: "dashboards"
+      path: "dashboard"
+  users_stream:
```
Comment (davydov-d) [inline airbyte-integrations/connectors/source-metabase/source_metabase/metabase.yaml:59]: does this stream have a primary key? it's missing here

**Review by davydov-d — APPROVED** [review]

File: airbyte-integrations/connectors/source-metabase/source_metabase/components.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        json={"username": username, "password": password},
+    )
+    response.raise_for_status()
+    if not response.ok:
+        raise ConnectionError(f"Failed to retrieve new session token, response code {response.status_code} because {response.reason}")
+    return response.json()["id"]
+
+
+@dataclass
+class MetabaseSessionTokenAuthenticator(AbstractHeaderAuthenticator, DeclarativeAuthenticator, JsonSchemaMixin):
```
[the remark this answers is not in the document]
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/source_metabase/components.py]: done, https://github.com/airbytehq/airbyte/pull/19716

File: airbyte-integrations/connectors/source-metabase/unit_tests/test_component.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
+#
```
Comment (girarda) [inline airbyte-integrations/connectors/source-metabase/unit_tests/test_component.py:1]: is this file still needed?

File: airbyte-integrations/connectors/source-metabase/setup.py [hunk 14]
```diff
@@ -5,7 +5,7 @@
 
 from setuptools import find_packages, setup
 
-MAIN_REQUIREMENTS = ["airbyte-cdk~=0.2", "requests>=2.28.0", "types-requests>=2.27.30"]
+MAIN_REQUIREMENTS = ["airbyte-cdk", "requests>=2.28.0", "types-requests>=2.27.30", "cachetools"]
```
Comment (girarda) [inline airbyte-integrations/connectors/source-metabase/setup.py:8]: we shouldn't need this dependency here since it'll be included in the cdk as part of https://github.com/airbytehq/airbyte/pull/19716

Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/setup.py]: thanks, removed

File: airbyte-integrations/connectors/source-metabase/unit_tests/test_component.py (hunk 13 above)
Comment (darynaishchenko) [reply airbyte-integrations/connectors/source-metabase/unit_tests/test_component.py]: removed

**Review by girarda — APPROVED** [review]
🚢

**Comment by lazebnyi** [comment]
/publish connector=connectors/source-metabase run-tests=false
